src/Microsoft.AspNetCore.Razor/CodeGenerators/CSharpTagHelperCodeRenderer.cs

```diff
@@ -450,112 +449,92 @@ protected virtual string GenerateUniqueId()
+                // writer.
+                var currentTargetWriter = _context.TargetWriterName;
+                var currentWriteAttributeMethodName = _context.Host.GeneratedClassContext.WriteAttributeValueMethodName;
+                _context.TargetWriterName = null;
 
-                    Debug.Assert(attributeValueChunk is ParentChunk);
-                    var children = ((ParentChunk)attributeValueChunk).Children;
-                    var attributeCount = children.Count(c => c is DynamicCodeAttributeChunk || c is LiteralCodeAttributeChunk);
+                Debug.Assert(attribute.Value is ParentChunk);
+                var children = ((ParentChunk)attribute.Value).Children;
+                var attributeCount = children.Count(c => c is DynamicCodeAttributeChunk || c is LiteralCodeAttributeChunk);
 
-                    _writer
-                        .WriteStartMethodInvocation(_tagHelperContext.BeginAddHtmlAttributeValuesMethodName)
-                        .Write(ExecutionContextVariableName)
-                        .WriteParameterSeparator()
-                        .WriteStringLiteral(attributeName)
-                        .WriteParameterSeparator()
-                        .Write(attributeCount.ToString(CultureInfo.InvariantCulture))
-                        .WriteEndMethodInvocation();
+                _writer
+                    .WriteStartMethodInvocation(_tagHelperContext.BeginAddHtmlAttributeValuesMethodName)
+                    .Write(ExecutionContextVariableName)
+                    .WriteParameterSeparator()
+                    .WriteStringLiteral(attribute.Name)
+                    .WriteParameterSeparator()
+                    .Write(attributeCount.ToString(CultureInfo.InvariantCulture))
+                    .WriteParameterSeparator()
+                    .Write(attributeValueStyleParameter)
+                    .WriteEndMethodInvocation();
 
-                    _attributeCodeVisitor.Accept(attributeValueChunk);
+                _attributeCodeVisitor.Accept(attribute.Value);
 
-                    _writer.WriteMethodInvocation(
-                        _tagHelperContext.EndAddHtmlAttributeValuesMethodName,
-                        ExecutionContextVariableName);
+                _writer.WriteMethodInvocation(
+                    _tagHelperContext.EndAddHtmlAttributeValuesMethodName,
+                    ExecutionContextVariableName);
 
-                    _context.TargetWriterName = currentTargetWriter;
-                }
-                else
-                {
-                    // HTML attributes are always strings. This attribute contains C# but is not dynamic. This occurs
-                    // when the attribute is a data-* attribute.
+                _context.TargetWriterName = currentTargetWriter;
+            }
+            else
+            {
+                // HTML attributes are always strings. This attribute contains C# but is not dynamic. This occurs
```

Comment:
"contains C# but is not dynamic"? This case is very similar to the `IsDynamicAttributeValue()` case and fits perfectly into the usual notion of "dynamic".

The difference is that attribute removal isn't done for `data-` attributes. Therefore I suggest something like `This is a data-* attribute which includes C#. Do not perform the conditional attribute removal or other special cases used when IsDynamicAttributeValue(). But the attribute must still be buffered to determine its final value.`